testnet/stacks-node/src/run_loop/boot_nakamoto.rs

```diff
@@ -0,0 +1,220 @@
+        if !exited_for_transition {
+            info!("Shutting down epoch-2/3 transition thread");
+            return;
+        }
+        info!("Reached Epoch-3.0 boundary, starting nakamoto node");
+        termination_switch.store(true, Ordering::SeqCst);
+        let naka = NakaRunLoop::new(
+            self.config.clone(),
+            Some(termination_switch),
+            Some(counters),
+        );
+        let new_coord_channels = naka
+            .get_coordinator_channel()
+            .expect("FATAL: should have coordinator channel in newly instantiated runloop");
+        {
+            let mut coord_channel = self.coordinator_channels.lock().expect("Mutex poisoned");
+            *coord_channel = new_coord_channels;
+        }
+        self.active_loop = InnerLoops::Epoch3(naka);
+        let InnerLoops::Epoch3(ref mut naka_loop) = self.active_loop else {
+            panic!("FATAL: unexpectedly found epoch2 loop after setting epoch3 active");
+        };
+        naka_loop.start(burnchain_opt, mine_start)
+    }
+
+    fn spawn_stopper(
+        config: &Config,
+        neon: &NeonRunLoop,
+    ) -> Result<JoinHandle<bool>, std::io::Error> {
+        let neon_term_switch = neon.get_termination_switch();
+        let config = config.clone();
+        thread::Builder::new()
+            .name("epoch-2/3-boot".into())
+            .spawn(move || {
+                loop {
+                    let do_transition = Self::reached_epoch_30_transition(&config)
+                        .unwrap_or_else(|err| {
+                            warn!("Error checking for Epoch-3.0 transition: {err:?}. Assuming transition did not occur yet.");
+                            false
+                        });
+                    if do_transition {
+                        break;
+                    }
+                    if !neon_term_switch.load(Ordering::SeqCst) {
+                        info!("Stop requested, exiting epoch-2/3-boot thread");
+                        return false;
+                    }
+                    thread::sleep(Duration::from_secs(1));
+                }
+                // if loop exited, do the transition
+                info!("Epoch-3.0 boundary reached, stopping Epoch-2.x run loop");
+                neon_term_switch.store(false, Ordering::SeqCst);
+                return true
+            })
+    }
+
+    fn reached_epoch_30_transition(config: &Config) -> Result<bool, String> {
+        let burn_height = Self::get_burn_height(config)?;
+        let epochs = StacksEpoch::get_epochs(
+            config.burnchain.get_bitcoin_network().1,
+            config.burnchain.epochs.as_ref(),
+        );
+        let epoch_3 = &epochs[StacksEpoch::find_epoch_by_id(&epochs, StacksEpochId::Epoch30)
+            .ok_or("No Epoch-3.0 defined")?];
+
+        Ok(u64::from(burn_height) >= epoch_3.start_height - 1)
+    }
+
+    fn get_burn_height(config: &Config) -> Result<u32, String> {
```

Comment:
Does this have to return a result? seems like it always returns Ok?

Reply:
Yeah, it's just to make it easier to return an error in the future if we want to error for any reason in this function (rather than defaulting to 0). One of the lessons for me from a lot of the interfaces in the `testnet/stacks-node` is that most functions should have been returning `Result`s.